File: template/examples/example_with_acd.c
/*
 * Example Source File with ACD Metadata
 * 
 * This file demonstrates how to add SCIS (Source Code Intelligence Standard)
 * metadata to your source code for ACD compliance.
 * 
 * Copyright (C) [Year] [Your Name/Organization]
 * Licensed under [Your License]
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Type definitions
typedef int error_code_t;
typedef struct context_t context_t;

// Error codes
#define SUCCESS 0
#define ERROR_INVALID_PARAM -1
#define ERROR_OUT_OF_MEMORY -2

/*
 * AI_PHASE: INIT
 * AI_STATUS: IMPLEMENTED
 * AI_COMPLEXITY: LOW
 * AI_NOTE: Initializes the application context with default values
 * AI_DEPENDENCIES: 
 * AI_COMMIT: initial
 */
error_code_t initialize_context(context_t** ctx) {
    if (ctx == NULL) {
        return ERROR_INVALID_PARAM;
    }
    
    *ctx = (context_t*)malloc(sizeof(context_t));
    if (*ctx == NULL) {
        return ERROR_OUT_OF_MEMORY;
    }
    
    // Initialize context members
    memset(*ctx, 0, sizeof(context_t));
    
    return SUCCESS;
}

/*
 * AI_PHASE: CLEANUP
 * AI_STATUS: IMPLEMENTED
 * AI_COMPLEXITY: LOW
 * AI_NOTE: Cleans up and frees the application context
 * AI_DEPENDENCIES: INIT
 * AI_COMMIT: initial
 */
error_code_t cleanup_context(context_t* ctx) {
    if (ctx == NULL) {
        return ERROR_INVALID_PARAM;
    }
    
    // Free any allocated resources
    free(ctx);
    
    return SUCCESS;
}

/*
 * AI_PHASE: ERROR_HANDLING
 * AI_STATUS: IMPLEMENTED
 * AI_COMPLEXITY: LOW
 * AI_NOTE: Converts error codes to human-readable strings
 * AI_DEPENDENCIES: 
 * AI_COMMIT: initial
 */
const char* error_to_string(error_code_t error) {
    switch (error) {
        case SUCCESS:
            return "Success";
        case ERROR_INVALID_PARAM:
            return "Invalid parameter";
        case ERROR_OUT_OF_MEMORY:
            return "Out of memory";
        default:
            return "Unknown error";
    }
}

/*
 * AI_PHASE: CORE_LOGIC
 * AI_STATUS: PARTIAL
 * AI_COMPLEXITY: MEDIUM
 * AI_NOTE: Main processing function - implementation in progress
 * AI_DEPENDENCIES: INIT, ERROR_HANDLING
 * AI_COMMIT: initial
 */
error_code_t process_data(context_t* ctx, const char* input, char** output) {
    if (ctx == NULL || input == NULL || output == NULL) {
        return ERROR_INVALID_PARAM;
    }
    
    // TODO: Implement data processing logic
    // This is a placeholder implementation
    
    *output = NULL;
    return SUCCESS;
}

/*
 * AI_PHASE: VALIDATION
 * AI_STATUS: IMPLEMENTED
 * AI_COMPLEXITY: LOW
 * AI_NOTE: Validates input parameters before processing
 * AI_DEPENDENCIES: ERROR_HANDLING
 * AI_COMMIT: initial
 */
error_code_t validate_input(const char* input) {
    if (input == NULL) {
        return ERROR_INVALID_PARAM;
    }
    
    if (strlen(input) == 0) {
        return ERROR_INVALID_PARAM;
    }
    
    return SUCCESS;
}

// Example main function
int main(int argc, char* argv[]) {
    context_t* ctx = NULL;
    error_code_t result;
    
    // Initialize
    result = initialize_context(&ctx);
    if (result != SUCCESS) {
        fprintf(stderr, "Initialization failed: %s\n", error_to_string(result));
        return 1;
    }
    
    printf("Application initialized successfully\n");
    
    // Cleanup
    result = cleanup_context(ctx);
    if (result != SUCCESS) {
        fprintf(stderr, "Cleanup failed: %s\n", error_to_string(result));
        return 1;
    }
    
    printf("Application terminated successfully\n");
    return 0;
}

File: template/examples/example_with_acd.py
"""
Example Python file with ACD metadata

This demonstrates how to add SCIS metadata to Python code using docstrings.
"""

# AI_PHASE: INIT
# AI_STATUS: IMPLEMENTED
# AI_COMPLEXITY: LOW
# AI_NOTE: Initialize application configuration
# AI_DEPENDENCIES: 
# AI_COMMIT: initial
def initialize_config(config_file=None):
    """
    Initialize application configuration.
    
    Args:
        config_file: Optional path to configuration file
        
    Returns:
        dict: Configuration dictionary
    """
    config = {
        'debug': False,
        'log_level': 'INFO',
        'max_retries': 3
    }
    
    if config_file:
        # Load from file
        pass
    
    return config


# AI_PHASE: ERROR_HANDLING
# AI_STATUS: IMPLEMENTED
# AI_COMPLEXITY: LOW
# AI_NOTE: Custom exception class for application errors
# AI_DEPENDENCIES: 
# AI_COMMIT: initial
class ApplicationError(Exception):
    """Custom exception for application errors."""
    
    def __init__(self, message, error_code=None):
        super().__init__(message)
        self.error_code = error_code


# AI_PHASE: VALIDATION
# AI_STATUS: IMPLEMENTED
# AI_COMPLEXITY: MEDIUM
# AI_NOTE: Validates input data with type checking
# AI_DEPENDENCIES: ERROR_HANDLING
# AI_COMMIT: initial
def validate_input(data):
    """
    Validate input data.
    
    Args:
        data: Input data to validate
        
    Returns:
        bool: True if valid
        
    Raises:
        ApplicationError: If validation fails
    """
    if data is None:
        raise ApplicationError("Input data cannot be None", error_code="INVALID_INPUT")
    
    if not isinstance(data, (str, dict, list)):
        raise ApplicationError("Unsupported data type", error_code="TYPE_ERROR")
    
    return True


# AI_PHASE: CORE_LOGIC
# AI_STATUS: PARTIAL
# AI_COMPLEXITY: HIGH
# AI_NOTE: Main data processing function - implementation in progress
# AI_DEPENDENCIES: INIT, VALIDATION, ERROR_HANDLING
# AI_COMMIT: initial
def process_data(data, config=None):
    """
    Process input data.
    
    Args:
        data: Input data to process
        config: Optional configuration dictionary
        
    Returns:
        Processed data
        
    Raises:
        ApplicationError: If processing fails
    """
    if config is None:
        config = initialize_config()
    
    # Validate input
    validate_input(data)
    
    # TODO: Implement actual processing logic
    result = data
    
    return result


# AI_PHASE: LOGGING
# AI_STATUS: IMPLEMENTED
# AI_COMPLEXITY: LOW
# AI_NOTE: Simple logging utility
# AI_DEPENDENCIES: INIT
# AI_COMMIT: initial
def log_message(message, level='INFO'):
    """
    Log a message.
    
    Args:
        message: Message to log
        level: Log level (INFO, WARNING, ERROR)
    """
    print(f"[{level}] {message}")


if __name__ == '__main__':
    # Example usage
    config = initialize_config()
    log_message("Application started")
    
    try:
        result = process_data("test data", config)
        log_message(f"Processing complete: {result}")
    except ApplicationError as e:
        log_message(f"Error: {e} (code: {e.error_code})", level='ERROR')
